File: java/src/main/java/org/libpetri/smt/encoding/IncidenceMatrix.java
package org.libpetri.smt.encoding;

/**
 * Incidence matrix for a flattened Petri net.
 *
 * <p>The incidence matrix C is defined as C[t][p] = post[t][p] - pre[t][p].
 * It captures the net effect of each transition on each place.
 *
 * <p>P-invariants are solutions to y^T * C = 0, found via null space
 * computation on C^T.
 */
public final class IncidenceMatrix {

    private final int[][] pre;      // pre[t][p]
    private final int[][] post;     // post[t][p]
    private final int[][] incidence; // C[t][p] = post - pre
    private final int numTransitions;
    private final int numPlaces;

    private IncidenceMatrix(int[][] pre, int[][] post, int[][] incidence, int numTransitions, int numPlaces) {
        this.pre = pre;
        this.post = post;
        this.incidence = incidence;
        this.numTransitions = numTransitions;
        this.numPlaces = numPlaces;
    }

    /**
     * Computes the incidence matrix from a FlatNet.
     *
     * <p>Environment-injected places (VER-006) each contribute one extra
     * <b>injector column</b> (a virtual transition that produces one token into
     * that place and consumes nothing). This makes P-invariant computation
     * env-aware: a valid invariant {@code y} must satisfy {@code y^T·C = 0} for the
     * injector column too, forcing {@code y[envPlace] = 0} and thereby discarding
     * closed-net conservation laws (e.g. {@code IN + OUT = const}) that would
     * otherwise vacuously bound an injectable place.
     *
     * @param flatNet the flattened net
     * @return the incidence matrix
     */
    public static IncidenceMatrix from(FlatNet flatNet) {
        int T = flatNet.transitionCount();
        int P = flatNet.placeCount();

        // Resolve injector columns: one per injected env place present in the index.
        var injectorPlaceIndices = new java.util.ArrayList<Integer>();
        for (var place : flatNet.environmentInjection().keySet()) {
            int idx = flatNet.indexOf(place);
            if (idx >= 0) injectorPlaceIndices.add(idx);
        }
        int E = injectorPlaceIndices.size();
        int rows = T + E;

        int[][] pre = new int[rows][P];
        int[][] post = new int[rows][P];
        int[][] incidence = new int[rows][P];

        for (int t = 0; t < T; t++) {
            var ft = flatNet.transitions().get(t);
            for (int p = 0; p < P; p++) {
                pre[t][p] = ft.preVector()[p];
                post[t][p] = ft.postVector()[p];
                incidence[t][p] = post[t][p] - pre[t][p];
            }
        }

        // Injector rows: pre = 0, post = e_p, incidence = e_p.
        for (int e = 0; e < E; e++) {
            int pid = injectorPlaceIndices.get(e);
            post[T + e][pid] = 1;
            incidence[T + e][pid] = 1;
        }

        return new IncidenceMatrix(pre, post, incidence, rows, P);
    }

    /**
     * Returns C^T (transpose of incidence matrix), dimensions [P][T].
     * Used for P-invariant computation: null space of C^T gives P-invariants.
     */
    public int[][] transposedIncidence() {
        int[][] ct = new int[numPlaces][numTransitions];
        for (int t = 0; t < numTransitions; t++) {
            for (int p = 0; p < numPlaces; p++) {
                ct[p][t] = incidence[t][p];
            }
        }
        return ct;
    }

    /**
     * Returns the pre-matrix (tokens consumed).
     * Shared reference — do not mutate.
     */
    public int[][] pre() { return pre; }

    /**
     * Returns the post-matrix (tokens produced).
     * Shared reference — do not mutate.
     */
    public int[][] post() { return post; }

    /**
     * Returns the incidence matrix C[t][p] = post - pre.
     * Shared reference — do not mutate.
     */
    public int[][] incidence() { return incidence; }

    public int numTransitions() { return numTransitions; }

    public int numPlaces() { return numPlaces; }
}

File: typescript/src/verification/encoding/incidence-matrix.ts
import type { FlatNet } from './flat-net.js';

/**
 * Incidence matrix for a flattened Petri net.
 *
 * The incidence matrix C is defined as C[t][p] = post[t][p] - pre[t][p].
 * It captures the net effect of each transition on each place.
 *
 * P-invariants are solutions to y^T * C = 0, found via null space
 * computation on C^T.
 */
export class IncidenceMatrix {
  private readonly _pre: readonly (readonly number[])[];
  private readonly _post: readonly (readonly number[])[];
  private readonly _incidence: readonly (readonly number[])[];
  private readonly _numTransitions: number;
  private readonly _numPlaces: number;

  private constructor(
    pre: number[][],
    post: number[][],
    incidence: number[][],
    numTransitions: number,
    numPlaces: number,
  ) {
    this._pre = pre;
    this._post = post;
    this._incidence = incidence;
    this._numTransitions = numTransitions;
    this._numPlaces = numPlaces;
  }

  /**
   * Computes the incidence matrix from a FlatNet.
   *
   * Environment-injected places (VER-006) each contribute one extra **injector
   * column** (a virtual transition that produces one token into that place and
   * consumes nothing). This makes P-invariant computation env-aware: a valid
   * invariant `y` must satisfy `y^T·C = 0` for the injector column too, forcing
   * `y[envPlace] = 0` and thereby discarding closed-net conservation laws (e.g.
   * `IN + OUT = const`) that would otherwise vacuously bound an injectable place.
   */
  static from(flatNet: FlatNet): IncidenceMatrix {
    const T = flatNet.transitions.length;
    const P = flatNet.places.length;

    const pre: number[][] = [];
    const post: number[][] = [];
    const incidence: number[][] = [];

    for (let t = 0; t < T; t++) {
      const ft = flatNet.transitions[t]!;
      const preRow = new Array<number>(P);
      const postRow = new Array<number>(P);
      const incRow = new Array<number>(P);

      for (let p = 0; p < P; p++) {
        preRow[p] = ft.preVector[p]!;
        postRow[p] = ft.postVector[p]!;
        incRow[p] = postRow[p]! - preRow[p]!;
      }

      pre.push(preRow);
      post.push(postRow);
      incidence.push(incRow);
    }

    // Injector columns (one per injected environment place): pre = 0, post = e_p.
    let injectorCount = 0;
    for (const name of flatNet.environmentInjection.keys()) {
      const idx = flatNet.placeIndex.get(name);
      if (idx == null) continue;
      const preRow = new Array<number>(P).fill(0);
      const postRow = new Array<number>(P).fill(0);
      const incRow = new Array<number>(P).fill(0);
      postRow[idx] = 1;
      incRow[idx] = 1;
      pre.push(preRow);
      post.push(postRow);
      incidence.push(incRow);
      injectorCount++;
    }

    return new IncidenceMatrix(pre, post, incidence, T + injectorCount, P);
  }

  /**
   * Returns C^T (transpose of incidence matrix), dimensions [P][T].
   * Used for P-invariant computation: null space of C^T gives P-invariants.
   */
  transposedIncidence(): number[][] {
    const ct: number[][] = [];
    for (let p = 0; p < this._numPlaces; p++) {
      const row = new Array<number>(this._numTransitions);
      for (let t = 0; t < this._numTransitions; t++) {
        row[t] = this._incidence[t]![p]!;
      }
      ct.push(row);
    }
    return ct;
  }

  /** Returns the pre-matrix (tokens consumed). T×P. */
  pre(): readonly (readonly number[])[] { return this._pre; }

  /** Returns the post-matrix (tokens produced). T×P. */
  post(): readonly (readonly number[])[] { return this._post; }

  /** Returns the incidence matrix C[t][p] = post - pre. T×P. */
  incidence(): readonly (readonly number[])[] { return this._incidence; }

  numTransitions(): number { return this._numTransitions; }
  numPlaces(): number { return this._numPlaces; }
}
